### DISCUSSION TITLE
How to switch cf2.0 to cf2.1

### DISCUSSION THREAD

**Discussion by CasperLiZ** [bitcraze/discussions#843] [Q&A] [answered]
I made a large-size 8520 hollow cup aircraft according to the CF2.1 schematic, but the solid information was displayed as CF2.0, which caused the IMU initialization to fail.
![image](https://github.com/bitcraze/discussions/assets/51102143/30db71f7-c14e-4726-ae1b-0e8fc5dba065)
Thanks

**Comment by krichardsson** [comment, marked as answer]
Hi,
there is something called the "device type", this is an identifier that is written to the hardware in production and is used when the system is booting to figure out how to configure it. If no device type identifier is found, the system will use Crazyflie 2.0 ("CF20") as a default, I assume this is what is happening in your case.

The device type is [read in platform.c](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/platform/src/platform.c#L72-L95).

There are two options for you
1. Write the correct device identifier to the hardware. I don't remember how to do this though, @tobbeanton can you help with this? Also, be a bit careful, I think this is memory that can only be written one time and never erased. 
2. Compile the firmware with `DEVICE_TYPE_STRING_FORCE` defined as "CF21" to override the default behavior. This is obviously easier and safer than 1.

**Reply by CasperLiZ** [reply]
Hi，
Thank you very much for your reply, it is my honor！I will try to modify "DEVICE_TYPE_STRING_FORCE", obviously this is the easiest and safest way, the first way I found the OTP address in the source code, but after looking up the information, I feel that it is relatively adventure. Finally, thank you again for your reply, best wishes.

**Reply by justinleeyang** [reply]
how to write it use otp? @tobbeanton @ataffanel @krichardsson

**Reply by CasperLiZ** [reply]
Using OTP is obviously very dangerous, because the number of times the STM32 OTP can be modified is limited. You can refer to
[How to write STM32F4 OTP Area](https://electronics.stackexchange.com/questions/423448/how-to-write-stm32f4-otp-area)

**Reply by CasperLiZ** [reply]
Using OTP is obviously very dangerous, because the number of times the STM32 OTP can be modified is limited. You can refer to
[How to write STM32F4 OTP Area](https://electronics.stackexchange.com/questions/423448/how-to-write-stm32f4-otp-area)

**Reply by justinleeyang** [reply]
if we compile the firmware with 2.1, and download , It is OK, but when cf-client-python OTA the HW, the firmware still be 2.0, how to slove the issue?

![image](https://github.com/user-attachments/assets/f1f5b212-a851-4020-81b2-a2b6988b5cd8)

**Reply by CasperLiZ** [reply]
You can see this instruction in the [src/platform/src/platform_stm32f4.c](https://github.com/bitcraze/crazyflie-firmware/blob/2cbf933c514b34427e1b2f531dc2d2f93dfb4938/src/platform/src/platform_stm32f4.c#L38) file. You only need to change the corresponding hardware version to achieve it. When using cf-client-python OTA, the firmware is still "DEFAULT_PLATFORM_STRING "0;CF20""

**Reply by justinleeyang** [reply]
I have tested ok use "0;CF21", but if cf-client-python OTA, still "DEFAULT_PLATFORM_STRING "0;CF20", crazyflie2.1 report MPU9250 error, no use BMI088, it is fail, how to sloved the issue ? only a method, OTP wirte ？

**Reply by justinleeyang** [reply]
I haved write OTP value as follow, but restart stm32f405 don't start up, why ?
![image](https://github.com/user-attachments/assets/a7885e99-7f46-4bd3-9ab5-0cdb2916e80b)
